tests/wake-format/basic/stdout

```diff
@@ -701,9 +702,13 @@ def loop aaaaa sssss eeeee =
 # A dictionary associating a key-value pair, providing fast lookup by key.
 # This is internally implemented by a balanced tree, so some total ordering must
 # be able to be produced for the key type.
+
```

Comment:
There's definitely arguments both ways, but if we've got some blank-line count we're enforcing between top-level `def`s (currently two, but I don't know if that's deliberate or just an artifact from something at the end of the blocks which is meant to eventually be fixed) I'd expect that same number of blank lines to be applied between "floating" top-level comments and ones more obviously attached to their token.  Then again, I'm not sure how you'd handle things like the below if you do go that way, especially with any consistency within blocks.  Maybe just enforce `^\s*#\s*$` as being the only way to connect distinct comment paragraphs, and otherwise they're treated as separate with whatever spacing is currently active (but minimum one)?
```wake
# A dictionary associating a key-value pair, providing fast lookup by key.
# This is internally implemented by a balanced tree, so some total ordering must
# be able to be produced for the key type.

# Second paragraph of the same documentation comment.
global export tuple Map k v =
```

Reply:
> but if we've got some blank-line count we're enforcing between top-level defs 

ATM it's intentionally forcing 2 based on the style guide doc, but its easily changeable

> I'd expect that same number of blank lines to be applied between "floating" top-level comments and ones more obviously attached to their token.

Unfortunately there isn't a method to know if a comment is floating vs bound. We could make some rule like "any comment surrounded by newlines" is floating and then we could force whatever rules about spacing. That has the tradeoff that authors can no longer space comments with newlines and must instead space them with empty comments

```
# one
# 
# two
# three
def x = y
```
vs 
```
# one
 
# two
# three
def x = y
```

I'm not opposed to this, but we will have to make a decision now. We can't punt/bikeshed because its a foundational/API level decision.

I'm mildly partial to the thing that doesn't require a rewrite but this is a now or never kind of thing.

Reply:
Though the more I think about it, the more I think floating comments are the right solution.

Let me sleep on it and see how I feel tomorrow morning

Reply:
It's also probably something worth asking at the Wake Dev meeting, which aligns nicely with sleeping on it.

Reply:
hmm. I thought of a tricky case here. Top level floating comments are fine but weird nested ones aren't so nice.

```
def thing =  (
    a
    b
    c

    # comment

    d
    e
    f
)
```

is very problematic. That `# comment` really needs to be a bound comment otherwise we are back the the core problem of checking for comments on every line of the formatter. So maybe we could change the rule to `any *top level* comment surrounded by newlines is floating` but that is a much harder thing to tag

Reply:
Ah, yeah, that's not a friendly case.  We've already got an output difference between top-level and internal tokens in `def` spacing so *conceptually* I don't think treating top comments differently would be too jarring, but that definitely doesn't help anything on the implementation side.